orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,156 @@ class PipelineSplitter {
   }
 };
 
+// TODO: move to a proper location for pipeline training
+struct PipelineBatchInfo {
+  std::vector<std::pair<int64_t, int64_t>> events;
+  int64_t retired_batch;  // index of retired batch, so its data could be reused
+};
+
+class PipelineBatchPlanner {
+ private:
+  int64_t max_id_;
+
+ public:
+  PipelineBatchPlanner()
+      : max_id_(::onnxruntime::contrib::OrtEventPool::GetPoolSize() - 1) {
+  }
+
+  void CreatePlan(const std::vector<int64_t>& start_ids,
+                  size_t num_batches,
+                  std::vector<PipelineBatchInfo>& plan) {
+    size_t num_stages = start_ids.size();
+
+    // generate timeline
+    Timeline timeline = GenerateTimeline(num_stages, num_batches);
+
+    // fill in plan
+    plan.resize(num_batches);
+    for (size_t batch = 0; batch < num_batches; ++batch) {
+      auto& info = plan[batch];
+      info.events.resize(2 * num_stages - 1);  // two pairs for each stage except for last stage, which only need one
+      info.retired_batch = -1;
+    }
+
+    // each stage requires 2 pair of wait/record events for FW/BW
+    // except for last stage, which only need one pair of wait/record
+    for (size_t s = 0; s < num_stages; ++s) {
+      int64_t prev_event_id = -1;
+      int64_t event_id = start_ids[s];
+      int64_t retired_batch = -1;
+      for (size_t t = 0; t < timeline.GetNumSlots(); ++t) {
+        const auto& slot = timeline.Get(s, t);
+        if (!slot.occupied)
+          continue;
+
+        ORT_ENFORCE(event_id < max_id_);
+        if (slot.is_backward) {
+          if (s < num_stages - 1) {
+            // dont update wait event for last stage's bw
+            plan[slot.batch_id].events[2 * num_stages - 2 - s].first = prev_event_id;
+            prev_event_id = event_id;
+          }
+          plan[slot.batch_id].events[2 * num_stages - 2 - s].second = event_id++;
+          retired_batch = gsl::narrow<int64_t>(slot.batch_id);
+        } else {
+          if (s == 0) {
+            // update retired batch when starting a new batch (s == 0 && !bw)
+            plan[slot.batch_id].retired_batch = retired_batch;
+          }
+          plan[slot.batch_id].events[s].first = prev_event_id;
+          prev_event_id = event_id;
+          if (s < num_stages - 1) {
+            // dont update record event for last stage's fw
+            plan[slot.batch_id].events[s].second = event_id++;
+          }
+        }
+      }
+    }
+  }
+
+ private:
+  class Timeline {
+   public:
+    struct Slot {
+      size_t batch_id;
+      bool is_backward;
```

Comment:
>bool is_backward [](start = 6, length = 16)

Consider make this a enum type 
WorkType work_type;

For now, WorkType only have forward and backward... but I feel this field would extend with fancier algorithm...